linux/HugePageMeter.c

```diff
@@ -0,0 +1,63 @@
+/*
+htop - HugePageMeter.c
+(C) 2021 htop dev team
+Released under the GNU GPLv2, see the COPYING file
+in the source distribution for its full text.
+*/
+
+#include "HugePageMeter.h"
+
+#include "LinuxProcessList.h"
+
+#include "CRT.h"
+#include "Object.h"
+#include "RichString.h"
+
+
+static const int HugePageMeter_attributes[] = {
+   MEMORY_USED,
+};
+
+static void HugePageMeter_updateValues(Meter* this, char* buffer, size_t size) {
+   int written;
+
+   const LinuxProcessList* lpl = (const LinuxProcessList*) this->pl;
+   this->total = lpl->totalHugePageMem;
+   this->values[0] = lpl->totalHugePageMem - lpl->freeHugePageMem;
+
+   written = Meter_humanUnit(buffer, this->values[0], size);
+   METER_BUFFER_CHECK(buffer, size, written);
+
+   METER_BUFFER_APPEND_CHR(buffer, size, '/');
+
+   Meter_humanUnit(buffer, this->total, size);
+}
+
+static void HugePageMeter_display(const Object* cast, RichString* out) {
+   char buffer[50];
+   const Meter* this = (const Meter*)cast;
+
+   RichString_writeAscii(out, CRT_colors[METER_TEXT], ":");
+   Meter_humanUnit(buffer, this->total, sizeof(buffer));
+   RichString_appendAscii(out, CRT_colors[METER_VALUE], buffer);
+
+   RichString_appendAscii(out, CRT_colors[METER_TEXT], " used:");
+   Meter_humanUnit(buffer, this->values[0], sizeof(buffer));
+   RichString_appendAscii(out, CRT_colors[MEMORY_USED], buffer);
+}
+
+const MeterClass HugePageMeter_class = {
+   .super = {
+      .extends = Class(Meter),
+      .delete = Meter_delete,
+      .display = HugePageMeter_display,
+   },
+   .updateValues = HugePageMeter_updateValues,
+   .defaultMode = BAR_METERMODE,
+   .maxItems = 1,
+   .total = 100.0,
+   .attributes = HugePageMeter_attributes,
+   .name = "HugePages",
+   .uiName = "HugePages",
+   .caption = "HuP"
```

Comment:
Just `HP ` maybe?

Reply:
For whatever it may be worth, I still think that people that use huge pages will want a meter per page size. So for example, x86-64 supports 2 MiB and 1 GiB huge pages. On other architectures, the available huge page sizes aren't friendly to two or three character abbreviations. For example, on arm64, a 512 MiB huge page size is possible: https://www.kernel.org/doc/html/latest/arm64/hugetlbpage.html

Reply:
Sadly Meters do not (yet) support options, so currently we would have to create (at compile time) Meters for all possible/supported page sizes.
As long we support *only* 2, 512 and 1024 MiB that might be a workaround.

Reply:
I was thinking about this some more. We could just have different values for "the" huge page meter, so rather than "used" versus "total", it becomes "used-2m", "used-1g", "total". And the most number of huge page sizes I've seen supported at the same time seems to be four. What do you think?

Reply:
We already support for meters to show more than just one value. So this probably won't be the main issue to show them. What might be somewhat harder, when not falling back to hardcoding the actual sizes is to transmit the actual labels. If the size of hugepages are more or less known, you could also just alloc an array with say 24 entries (1MiB to 8TiB) and just put the proper values in the right slot or `NAN` if unsupported. On the Meter side, coloring would than just count the used slots with the first used getting color 1, the second color 2, and so on. Sure some more LOC, but wouldn't limit extensibility too much and avoids hardcoding the labels (HugePages that ain't Power of Two make little sense).

Reply:
So think I've more or less implemented this model. Please see the following commit that builds on this pull request: https://github.com/davezarzycki/htop/commit/410c24341deaceb1b539fc8cad3873b0fb813eef

Reply:
Thanks for the feedback. Here is my updated branch that is derived from this pull request: https://github.com/davezarzycki/htop/tree/per_hugepage_values

Reply:
The above branch now has a fix for the hasty `nextUsed` change I just made.

Reply:
I think I've addressed all of the feedback to date, including the import order change that I overlooked yesterday.

I've also updated the branch after two additional tests where zero or one huge page size is available (rather than the two sizes on my system). Now zero or one available huge page size will print " used:" instead of the page size, and zero available huge page sizes are now treated as if one huge page size exists (with zero used and zero total).